### ISSUE TITLE
make depend strangeness

### ISSUE THREAD

**Issue by tessus** [openssl/openssl#770]
openssl 1.0.1g

When I run `./config -fPIC`, I get the following message at the end:

```
[snip]
Configured for linux-x86_64.

*** Because of configuration changes, you MUST do the following before
*** building:

        make depend
```

But it compiles and installs just fine with `make && make install`. There's also nothing in the INSTALL file that would explain this `make depend` change. Can you please shed some light on this?

**Comment by levitte** [comment]
That message is a bit overzealous at times.  It's worded a bit too strongly, see it as a reminder of something you might want to do.

**Comment by tessus** [comment]
Well, do I have to run it or not? If yes, why does everything work as expected when not doing it?
If no, why would I want to do it?

I just don't understand the reasoning and what it should accomplish. Let's put it that way: `make depend` sounds like it builds and resolves dependencies. But a make file would do that automatically in the first place. And since everythings works w/o running `make depend`, this makes no sense. For testing purposes I ran `make depend` and this took longer than the `make`. After the `make` I ran `make depend` again and it was starting over. Shouldn't it tell me 'nothing to do`?

Something is very strange and without proper documentation I don't know what's really going on.

**Comment by levitte** [comment]
`make depend` is a fairly old idiom for finding out all extra dependencies that are needed, by scanning all source files and find out what they include and generate the extra dependencies from those.  `make` has no way of doing that work on its own.  The GNU method, however, has automated this by generating those extra dependencies on the fly, so if you've grown up in a GNU world, `make depend` is certainly a very strange concept to start with.

If your doing a `make depend` generated a rebuild, it's because `make` has new dependency information and some of it triggered a rebuild.  This may happen after a `make depend`, but only once.

If your build went through fine and tested fine, you're fine.

**Comment by tessus** [comment]
Thank you very much for the explanation.

How do I know, if all tests passed? e.g. I saw some ocsp verify errors, but I do not know, if that was expected or not, because it doesn't tell me, if the test passed or not.
The error could mean the test passed as well as that it failed.

```
DELEGATED; Root CA -> EE
Response Verify Failure
140497442698912:error:0407006A:rsa routines:RSA_padding_check_PKCS1_type_1:block type is not 01:rsa_pk1.c:103:
140497442698912:error:04067072:rsa routines:RSA_EAY_PUBLIC_DECRYPT:padding check failed:rsa_eay.c:705:
140497442698912:error:0D0C5006:asn1 encoding routines:ASN1_item_verify:EVP lib:a_verify.c:218:
140497442698912:error:27069065:OCSP routines:OCSP_basic_verify:certificate verify error:ocsp_vfy.c:138:Verify error:certificate sign
ature failure
=== INVALID SIGNATURE on the ISSUER CERTIFICATE ===
NON-DELEGATED; Intermediate CA -> EE
Response verify OK
```

**Comment by levitte** [comment]
That particular series of OCSP tests starts with this title:

    === INVALID SIGNATURE on the DELEGATED OCSP SIGNING CERTIFICATE ===

That's what those tests do.

I assume the OCSP tests also ended like this?

    ALL OCSP TESTS SUCCESSFUL

I will agree with you, though, that the test report isn't the best in the world.  We've changed that for OpenSSL 1.1.  The 1.0.2 series is in maintainance mode, changing the tests in some radical way is unfortunately out of scope.

**Comment by tessus** [comment]
Yes, I can see the `ALL OCSP TESTS SUCCESSFUL`, but there are a lot more tests above that one.
I haven't counted the lines but the output is really long, so what shall I do?
Parse for `ALL * TESTS SUCCESSFUL`? What would be the message, if there was a failed test?

Something like `grep "TESTS SUCCESSFUL" output.txt |wc -l` would do it, if I knew how many there should be.

**Comment by levitte** [comment]
All tests will exit with an exit code other than 0, which will have `make` simply halt.  That'll be your indication that a test went wrong (and which one, because it'll be the last one that was performed).

A successful test suite in 1.0.x will end with a call of `openssl version -a`, which gives an output similar to this:

```
OPENSSL_CONF=apps/openssl.cnf util/opensslwrap.sh version -a
OpenSSL 1.0.2g-dev  xx XXX xxxx
built on: reproducible build, date unspecified
platform: linux-x86_64
options:  bn(64,64) rc4(16x,int) des(idx,cisc,16,int) idea(int) blowfish(idx) 
compiler: gcc -I. -I.. -I../include  -DOPENSSL_THREADS -D_REENTRANT -DDSO_DLFCN -DHAVE_DLFCN_H -Wa,--noexecstack -m64 -DL_ENDIAN -O3 -Wall -DOPENSSL_IA32_SSE2 -DOPENSSL_BN_ASM_MONT -DOPENSSL_BN_ASM_MONT5 -DOPENSSL_BN_ASM_GF2m -DSHA1_ASM -DSHA256_ASM -DSHA512_ASM -DMD5_ASM -DAES_ASM -DVPAES_ASM -DBSAES_ASM -DWHIRLPOOL_ASM -DGHASH_ASM -DECP_NISTZ256_ASM
OPENSSLDIR: "/usr/local/ssl"
```

The details will differ, such as the date (mine is `xx XXX xxxx` since I run the development version) and other stuff, but basically that.

**Comment by tessus** [comment]
`make` abending with a retun code other than 0 is fine. e.g. I also didn't know that the test suite would just stop. This answers all my questions.

Thanks again for your explanation and help.
